protocol/network/2-state-of-the-art.md

```diff
@@ -0,0 +1,70 @@
+2 An analysis the State of the Art in Network Stacks
+====================================================
+
+This section presents to the reader an analysis of the available protocols and arquitectures for a Network Stack. The goal is to provide the foundations from which to infer the conclusions and understand what are libp2p requirements and its designed arquitecture.
+
+## 2.1 The client-server model
+
+The client-server model indicates that both parties that ends of the channel have different roles, that support different services and/or have different capabilities, or in another words, speak different protocols.
+
+Building client-server applications has been natural tendency for a number of reasons:
+
+- The bandwidth inside a DataCenter is considerably high compared to the one available for clients to connect between each other
+- DataCenter resources are considerably cheaper, due to efficient usage and bulk stocking
+- Enables easier methods for the developer and system admin to have a fine grained control over the application
+- Reduces the number of heteregeneus systems to be handled (although it is still considerably high)
+- Systems like NAT make it really hard for client machines to find and talk with each other, forcing a developer to perform very clever hacks to traverse these obstacles.
+- Protocols started to be designed with the assumption that a developer will create a client-server application from the start.
+
+We even learned how to hide all of the complexity of a distributed system behind gateways on the Internet, using protocols that were designed to perform a point-to-point operation, such as HTTP, making it opaque for the application to see and understand how the cascade of service calls made for each request.
+
+`libp2p` offers a move towards dialer-listener interactions, from the client-server listener, where it is not implicit which of the entities, dialer or listener, has which capabilities or is enabled to perform which actions. Setting up a connection between two applications today is a multilayered problem to solve, and these connections should not have a purpose bias, instead support to several other protocols to work on top of the established connection. In a client-server model, a server sending data without a prior request from the client is known as a push model, which typically adds more complexity, in a dialer-listener model, both entities can perform requests independently.
+
+## 2.2 Categorizing the network stack protocols by solutions
```

Comment:
maybe a note describing why this is here:

> Before diving into the libp2p protocols, it is important to understand the large diversity of protocols already in wide use and deployment that help maintain today's simple abstractions. For example, when one thinks about an HTTP connection, one might naively just think `HTTP/TCP/IP` as the main protocols involved, but in reality many more participate, all depending on the usage, the networks involved, and so on. Protocols like `DNS, DHCP, ARP, OSPF, Ethernet, 802.11 (WiFI), ...` and many others get involved. Looking inside ISPs' own networks would reveal dozens more.
>
> Additionally, it's worth noting that the traditional 7-layer OSI model characterization does not fit libp2p, because the OSI model tends to drive towards point-to-point links between applications, whereas the libp2p protocols speak more towards various sizes of networks with various properties cooperating under various  different security models. Also, libp2p protocols can "peer at the same level", meaning that multiple protocols can run simultaneously to handle one type of responsibility, instead of one-per-layer in traditional OSI stacking (for example, `bootstrap lists`, `mDNS`, `DHT Discovery`, and `PEX` are all forms of "_Peer Discovery_" that can coexist and can synergize.